Zephyr main efr32bg22 support

sateeshkotapati opened the pull request:
Added support for the BRD4184B board, as part of efr32_radio.
Support for the EFR32BG22 SoC was added in main Zephyr repo to achieve this.
Please review and appreciate your time. Thanks


P.S: hal_silabs modules PR is https://github.com/zephyrproject-rtos/hal_silabs/pull/13

sateeshkotapati commented:
I see a failure with clang-build (4). It is complaining about not having enough memory. 
Is that correct or am I missing something? How to restart the check if it is only memory issue?

https://github.com/zephyrproject-rtos/zephyr/actions/runs/856837557#artifacts

chrta commented on subsys/pm/policy/Kconfig:
```diff
@@ -2,6 +2,7 @@
 
 choice
 	prompt "Idle State Power Management Policy"
+	default PM_POLICY_APP if SOC_SERIES_EFR32BG22
```
I do not think this is the correct location to do this.

sateeshkotapati replied:
removed

chrta commented on tests/drivers/i2c/i2c_gecko/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
I am not sure if this is a valid test that can be added to zephyr. If it is, the folder "tests/boards" would be the better location.

chrta commented on drivers/counter/counter_gecko_stimer.c:
```diff
@@ -0,0 +1,291 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT silabs_gecko_rtcc
```
I do not think that a different driver with the same compatibility string is a good idea. As far as i understand both drivers counter_gecko_rtcc.c and counter_gecko_stimer.c use the same hardware peripheral and it is just a different driver implementation.

sateeshkotapati replied:
Created a different compatibility string.

chrta commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.defconfig.efr32bg22:
```diff
@@ -0,0 +1,24 @@
+	depends on SPI
+	
+config COUNTER_GECKO_STIMER	
+	default y
```
Is this driver mandatory for the efr32bg22? With this line it is always compiled in (so if PM is enabled and if it is disabled).

sateeshkotapati replied:
This driver is required even if PM is not enabled.

chrta reviewed (changes requested):
Thank you for your contribution. This is a lot of new stuff. I am not sure if the manual addition of the Si7021 code in the samples and tests should be added.
Also i have not used power management, so i can only give some very basic comments there.

sateeshkotapati commented on boards/arm/efr32_radio/doc/brd4184b.rst:
```diff
@@ -0,0 +1,146 @@
+.. _efr32_radio_brd4184b:
```
Moved board efr32bg-sltb010a to a standalone folder.

sateeshkotapati commented on boards/arm/efr32_radio/doc/brd4184b.rst:
```diff
@@ -0,0 +1,146 @@
++-----------+------------+-------------------------------------+
+| SYSTICK   | on-chip    | systick                             |
++-----------+------------+-------------------------------------+
+| COUNTER   | on-chip    | rtcc                                |
```
It is stimer based on rtcc driver.

sateeshkotapati commented on boards/arm/efr32_radio/doc/brd4184b.rst:
```diff
@@ -0,0 +1,146 @@
+- `EFR32BG22 Datasheet`_
+- `EFR32xG22 Reference Manual`_
+- `BRD4184B User Guide`_
+
```
Done

sateeshkotapati commented on boards/arm/efr32_radio/efr32_radio_brd4184b.dts:
```diff
@@ -0,0 +1,95 @@
+	location-sda= <GECKO_LOCATION(0) GECKO_PORT_D GECKO_PIN(2)>;
+	location-scl= <GECKO_LOCATION(0) GECKO_PORT_D GECKO_PIN(3)>;
+	clock-frequency = <I2C_BITRATE_FAST>;
+	si7021:	si7021@40 {
```
Removed i2c and s7021 instances in this PR.

sateeshkotapati commented on boards/arm/efr32_radio/efr32_radio_brd4184b_defconfig:
```diff
@@ -0,0 +1,20 @@
+# Enable CONFIG_PM if the application needs power manager utilities
+CONFIG_PM=n
+
+
```
Done

sateeshkotapati commented on drivers/counter/Kconfig.gecko:
```diff
@@ -10,4 +10,4 @@ config COUNTER_GECKO_RTCC
 	help
 	  Enable counter driver based on RTCC module for Silicon Labs Gecko
-	  chips.
+	  chips.
```
Done

sateeshkotapati commented on drivers/entropy/entropy_gecko_trng.c:
```diff
@@ -13,6 +13,7 @@
 
 static void entropy_gecko_trng_read(uint8_t *output, size_t len)
 {
+#ifndef CONFIG_CRYPTO_ACC_GECKO_TRNG		
```
I have kept as is because of the following two reasons.
1. CONFIG_CRYPTO_ACC_GECKO_TRNG is defined if SOC_SERIES_EFR32BG22 is defined.
2. If a new board is added which uses TRNG hardware module and not CRYPTO_ACC hardware module, then this has to be changed.

chrta replied:
In that case the Kconfig setting should be promtless, so it is not visible to the user when executing menuconfig or guiconfig, see https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html?highlight=kconfig%20prompt#id1

sateeshkotapati commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,266 @@
+/ {
+	chosen {
+		zephyr,flash-controller = &msc;
+		zephyr,console = &usart1;
```
Moved accordingly.

sateeshkotapati commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,266 @@
+	
+	/* These aliases are provided for compatibility with samples */
+	aliases {
+		led0 = &led0;
```
Moved accordingly

sateeshkotapati commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,266 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+		cpu0: cpu@0 {
+		    device_type = "cpu";
```
Done

sateeshkotapati commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,266 @@
+			compatible = "zephyr,power-state";
+			power-state-name = "standby";
+			min-residency-us = < 50000 >;
+			phandle = < 0x5 >;
```
removed

sateeshkotapati commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,266 @@
+			gpios = <&gpiob 1 GPIO_ACTIVE_LOW>;
+			label = "User Push Button 0";
+		};
+	};
```
Moved accordingly.

sateeshkotapati commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,266 @@
+			clock-frequency = <I2C_BITRATE_STANDARD>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x4a010000 0x3044>;
```
Register space has lot of reserved memory to align it to 1k words.

sateeshkotapati commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,266 @@
+
+&wdog0 {
+	status = "okay";
+};
```
moved

sateeshkotapati commented on samples/drivers/counter/alarm/src/main.c:
```diff
@@ -27,6 +27,8 @@ struct counter_alarm_cfg alarm_cfg;
 #elif defined(CONFIG_COUNTER_ESP32)
 #define TIMER DT_LABEL(DT_NODELABEL(timer0))
+#elif defined(CONFIG_COUNTER_GECKO_RTCC) || defined(CONFIG_COUNTER_GECKO_STIMER)
+#define RTCC_TIMER DT_LABEL(DT_NODELABEL(rtcc0))
```
Unfortunately, there is a TIMER already defined in EFR32BG soc header files.

sateeshkotapati commented on soc/arm/silabs_exx32/Kconfig.defconfig:
```diff
@@ -10,3 +10,10 @@ config SOC_GECKO_EMU
+config PM
+	default n
+	select COUNTER
+endif
```
Moved to proper place. This is defined in Kconfig.defconfig.series in soc series directory.

sateeshkotapati commented on soc/arm/silabs_exx32/common/CMakeLists.txt:
```diff
@@ -3,3 +3,8 @@
 zephyr_sources_ifdef(CONFIG_PM soc_power.c)
+
+zephyr_library_include_directories(
+  ${ZEPHYR_BASE}../../modules/hal/silabs/gecko/service/power_manager/inc
```
removed

sateeshkotapati commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -8,6 +9,56 @@
 
 #include <logging/log.h>
+
+#ifdef CONFIG_SOC_SERIES_EFR32BG22
```
Done

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.defconfig.efr32bg22:
```diff
@@ -0,0 +1,24 @@
+	
+config COUNTER_GECKO_STIMER	
+	default y
+	
```
Done

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.series:
```diff
@@ -0,0 +1,28 @@
+	select HAS_POWER_STATE_SLEEP_1
+	select HAS_POWER_STATE_SLEEP_2
+	select HAS_POWER_STATE_SLEEP_3
+        select CPU_CORTEX_M_HAS_SYSTICK
```
Fixed

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.soc:
```diff
@@ -0,0 +1,19 @@
+if SOC_SERIES_EFR32BG22
+
+config SOC_PART_NUMBER_EFR32BG22C224F512IM40
+	bool "SOC_LPC55S16 M33"
```
fixed

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.series:
```diff
@@ -0,0 +1,28 @@
+	select SOC_GECKO_EMU
+	select SOC_GECKO_GPIO
+	select SOC_GECKO_TRNG
+	select SOC_GECKO_RTCC
```
Yes, above defines are not required for EFR32BG22

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.soc:
```diff
@@ -0,0 +1,19 @@
+	select CPU_HAS_FPU
+	select ARMV8_M_DSP
+	select ARM_TRUSTZONE_M
+	select CLOCK_CONTROL
```
Fixed this by moving all these to Kconfig.series as they are not specific to soc

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/soc_pinmap__.h:
```diff
@@ -0,0 +1,34 @@
+/*
```
Done

sateeshkotapati commented:
I have incorporated the review comments. 
Thank you

Nukersson commented on CODEOWNERS:
```diff
@@ -370,6 +371,7 @@
 /dts/arm/silabs/efm32pg1b*                @rdmeneze
 /dts/arm/silabs/efr32mg21*                @l-alfred
 /dts/arm/silabs/efr32fg13*                @yonsch
+/dts/arm/silabs/efr32bg22*                @sateeshkotapati
```
I don't think, that this is a good idea to add yourself as maintainer without additional approvement from main dev group. @galak?

sateeshkotapati replied:
How to get additional approval from main dev group. It gave "check patch error" when I don't add myself here.

Nukersson replied:
Please read this doc first to understand responsibilities: https://docs.zephyrproject.org/latest/development_process/project_roles.html 

But I would say, that in your case the root of the problem is somewhere else.

sateeshkotapati replied:
I will be maintaining the code for EFR32BG22 in future.

Nukersson commented on boards/arm/efr32bg_sltb010a/doc/index.rst:
```diff
@@ -0,0 +1,184 @@
+   :alt: EFR32BG-SLTB010A
+
+Hardware
+********
```
The number of '*' must match the length of the word.

sateeshkotapati replied:
It is matching I guess. Am I missing something?

Nukersson replied:
Looks not matching. Maybe just formatting issue from my app. Take a look on ci output. There are many mistakes.

sateeshkotapati replied:
could you please help me how to check CI output?

chrta replied:
The length of the header should be correct. Both line 17 and line 18 contain 8 characters.

sateeshkotapati replied:
Done

Nukersson commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,135 @@
+};
+
+&cpu0 {
+	clock-frequency = <76800000>;
```
Use macro to define clock either in KHZ or MHZ

sateeshkotapati replied:
Except I2C_BITRATE_STANDARD, I don't see cpu-frequency is using a macro. If we need to define a macro, we need to add it in dt-bindings include file.

Nukersson replied:
Oh. Sorry. I thought, it was clock configuration in the board. Actually cpu node does not belongs to board usually, but to soc instead. Take a look here: https://docs.zephyrproject.org/latest/guides/porting/board_porting.html

It could be that some work is already done for your soc.

Nukersson commented on drivers/counter/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ zephyr_library()
 zephyr_library_sources_ifdef(CONFIG_TIMER_TMR_CMSDK_APB         timer_tmr_cmsdk_apb.c)
 zephyr_library_sources_ifdef(CONFIG_TIMER_DTMR_CMSDK_APB        timer_dtmr_cmsdk_apb.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_GECKO_RTCC          counter_gecko_rtcc.c)
+zephyr_library_sources_ifdef(CONFIG_COUNTER_GECKO_STIMER        counter_gecko_stimer.c)
```
The driver should be added in separate commit as with the board.

sateeshkotapati replied:
Done

KozhinovAlexander reviewed (changes requested):
Please split your commit into multiple ones aimed on different sections: board, drivers, soc etc.

sateeshkotapati committed:
drivers: hwinfo: gecko: Fix RMU_RST build error for efr32bg22-sltb010a

This commit fixes the build issue in hwinfo_gecko.c file for EFR32BG22

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

povergoing commented on west.yml:
```diff
@@ -66,7 +66,7 @@ manifest:
       path: modules/hal/microchip
     - name: hal_silabs
-      revision: be39d4eebeddac6e18e9c0c3ba1b31ad1e82eaed
+      revision: pull/13/head
```
Personally, I don't understand this change.

sateeshkotapati replied:
As per the guidelines in https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-module, 13 in the example above indicated the pull request number submitted to the hal_silabs repository. Once the module changes are reviewed and merged, the revision is changed to the commit hash from the module repository.

povergoing replied:
Really thanks, something I don't know before:)

chrta commented on subsys/pm/policy/Kconfig:
```diff
@@ -1,6 +1,6 @@
 # SPDX-License-Identifier: Apache-2.0
 
-choice
+choice PM_POLICY
```
Please revert this change if it is not necessary.

sateeshkotapati replied:
We were not able to define PM_POLICY_APP (if it is not named choice) because of this issue https://github.com/zephyrproject-rtos/zephyr/issues/6948

ioannisg commented on soc/arm/silabs_exx32/efr32bg22/Kconfig.series:
```diff
@@ -0,0 +1,30 @@
+	select SOC_FAMILY_EXX32
+	select HAS_SILABS_GECKO
+	select HAS_SWO
+	select CPU_CORTEX_M_HAS_SYSTICK
```
This is not reuquired. It's added to the Cortex-M33 definition.

sateeshkotapati replied:
Removed CPU_CORTEX_M_HAS_SYSTICK

ioannisg reviewed (commented):
Looks fine to me, pls address a comment I left about CPU_CORTEX_M_HAS_SYSTICK, and the comment about removing the diff in pm subsystem

sateeshkotapati commented:
> Looks fine to me, pls address a comment I left about CPU_CORTEX_M_HAS_SYSTICK, and the comment about removing the diff in pm subsystem

We were not able to define PM_POLICY_APP (if it is non-named choice) because of this issue #6948

sateeshkotapati committed:
boards: Add support for Silicon Labs' EFR32BG22-SLTB010A board

This commit adds support for Silicon Labs EFR32BG22-SLTB010A
(a.k.a. Thunderboard EFR32BG22)

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
soc: silabs_exx32: Add support for SiLabs EFR32BG22 SoC

This commit adds support for Silicon Labs EFR32BG22 (Blue Gecko) SoC.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
soc: silabs_exx32: Add support for SiLabs EFR32BG22 SoC

This commit adds support for Silicon Labs EFR32BG22 (Blue Gecko) SoC.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
drivers: gecko: Add/modify support for EFR32BG22 drivers

This commit adds STIMER driver and modifies
TRNG and watchdog drivers for EFR32BG22.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
drivers: gecko: Add/modify support for EFR32BG22 drivers

This commit adds STIMER driver and modifies
TRNG and watchdog drivers for EFR32BG22.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
west.yml: Update hal_silabs module

Update west.yml to point to hal_silabs version with support for
Silicon Labs EFR32BG22 SoCs.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

chrta commented on soc/arm/silabs_exx32/Kconfig:
```diff
@@ -134,6 +134,27 @@ choice SOC_GECKO_EMU_DCDC_MODE
 
+config SOC_GECKO_DEV_INIT
+	bool
+	default n
```
The value n is the default in Kconfig, so this line can be removed (also line 146 and line 153 in this file)

chrta commented on drivers/hwinfo/hwinfo_gecko.c:
```diff
@@ -139,6 +149,9 @@ int z_impl_hwinfo_get_reset_cause(uint32_t *cause)
 
 #endif /* defined(BU_PRESENT) && defined(_SILICON_LABS_32B_SERIES_0) */
 
+	/* To fix compiler warning when none of the above flags are defined*/
+	rmu_flags = 0;
+
```
These 3 lines are redundant and should be removed. rmu_flags is always initialized in line 38 and rmu_flags is never read after this line.

sateeshkotapati replied:
Got a compilation warning as below:
C:/work/zephyrproject/zephyr/drivers/hwinfo/hwinfo_gecko.c:38:11: warning: unused variable 'rmu_flags' [-Wunused-variable]
   38 |  uint32_t rmu_flags = RMU_ResetCauseGet();
      |           ^~~~~~~~~

chrta replied:
OK, in that case using `ARG_UNUSED(rmu_flags)` is probably better. Just setting a variable without reading it may also create compiler warnings or issues with static code analysis tools.

chrta commented on boards/arm/efr32bg_sltb010a/doc/index.rst:
```diff
@@ -0,0 +1,184 @@
+- A physical UART connection which is relayed over interface USB Serial port.
+
+Flashing an application to EFR32BG-SLTB010A
+-----------------------------------------
```
The length of line 134 does not match the length of line 133, please add `--` to line 134

chrta commented on drivers/entropy/entropy_gecko_trng.c:
```diff
@@ -100,7 +121,7 @@ static struct entropy_driver_api entropy_gecko_trng_api_funcs = {
 };
 
 DEVICE_DT_INST_DEFINE(0,
-			entropy_gecko_trng_init, NULL,
-			NULL, NULL,
-			PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
-			&entropy_gecko_trng_api_funcs);
+				entropy_gecko_trng_init, NULL,
+				NULL, NULL,
+				PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+				&entropy_gecko_trng_api_funcs);
```
Please revert this change, you broke the indentation here.

chrta commented on boards/arm/efr32bg_sltb010a/doc/index.rst:
```diff
@@ -0,0 +1,184 @@
+Connections and IOs
+===================
+
+The EFR32BG SoC has eight gpio controllers (PORTA, PORTB, PORTC, PORTD,
```
eight -> six

sateeshkotapati replied:
changed

chrta commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,201 @@
+			label = "STIMER_0";
+		};
+
+		gpio: gpio@0x4003C000 {
```
Please remove the leading 0x from the node names. I get a some warnings, when i build, e.g. hello_world:

```
west build --pristine -b efr32bg_sltb010a samples/hello_world/
...
efr32bg_sltb010a.dts.pre.tmp:134.27-142.6: Warning (unit_address_format): /soc/gpio@0x4003C000/gpio@0x4003C000: unit name should not have leading "0x"
  also defined at efr32bg_sltb010a.dts.pre.tmp:286.8-288.3
efr32bg_sltb010a.dts.pre.tmp:143.27-151.6: Warning (unit_address_format): /soc/gpio@0x4003C000/gpio@0x4003C030: unit name should not have leading "0x"
  also defined at efr32bg_sltb010a.dts.pre.tmp:289.8-291.3
efr32bg_sltb010a.dts.pre.tmp:152.27-160.6: Warning (unit_address_format): /soc/gpio@0x4003C000/gpio@0x4003C060: unit name should not have leading "0x"
  also defined at efr32bg_sltb010a.dts.pre.tmp:292.8-294.3
efr32bg_sltb010a.dts.pre.tmp:161.27-169.6: Warning (unit_address_format): /soc/gpio@0x4003C000/gpio@0x4003C090: unit name should not have leading "0x"
  also defined at efr32bg_sltb010a.dts.pre.tmp:295.8-297.3
efr32bg_sltb010a.dts.pre.tmp:170.27-178.6: Warning (unit_address_format): /soc/gpio@0x4003C000/gpio@0x4003C0C0: unit name should not have leading "0x"
  also defined at efr32bg_sltb010a.dts.pre.tmp:298.8-300.3
efr32bg_sltb010a.dts.pre.tmp:67.26-75.5: Warning (simple_bus_reg): /soc/usart@4005C000: simple-bus unit address format error, expected "4005c000"
  also defined at efr32bg_sltb010a.dts.pre.tmp:239.9-244.3
efr32bg_sltb010a.dts.pre.tmp:85.23-91.5: Warning (simple_bus_reg): /soc/trng@4C021000: simple-bus unit address format error, expected "4c021000"
efr32bg_sltb010a.dts.pre.tmp:125.25-179.5: Warning (simple_bus_reg): /soc/gpio@0x4003C000: simple-bus unit address format error, expected "4003c000"
  also defined at efr32bg_sltb010a.dts.pre.tmp:282.7-285.3
efr32bg_sltb010a.dts.pre.tmp:180.24-187.5: Warning (simple_bus_reg): /soc/wdog@4A018000: simple-bus unit address format error, expected "4a018000"
  also defined at efr32bg_sltb010a.dts.pre.tmp:251.8-253.3
  also defined at efr32bg_sltb010a.dts.pre.tmp:301.8-303.3
...

```

sateeshkotapati replied:
Fixed

chrta commented on dts/arm/silabs/efr32bg22.dtsi:
```diff
@@ -0,0 +1,201 @@
+			};
+		};
+
+		usart0: usart@4005C000 { /* USART0 */
```
Please use lower-case letters in hex addresses in node names, otherwise an error is displayed, see other comment.

sateeshkotapati replied:
Done

chrta reviewed (changes requested):
Otherwise it looks good for my, just the 3 minor issues left.
Please make sure that no warnings are displayed when building the samples, e.g. samples/basic/blinky

chrta reviewed (dismissed):
From my point of view, this pr looks good. I am not sure about the PM changes. Someone with more knowledge in this area should check them.

galak commented:
Do we need all the service/ stuff as well?  I'd like to see if we can limit the initial import here to be similar to how all the other Silab SoCs are functioning.  Than we can see about adding this stuff in a follow on PR if that's feasible.

galak commented:
Sorry I didn't make this comment sooner, but can we try and limit the initial PR to the bare minimum to get the board and SoC going.  So whatever basic init code and uart functioning.  The other changes should than be split up into separate PRs.

For example, do we need the stimer driver for the board to boot and function?

sateeshkotapati commented:
> Sorry I didn't make this comment sooner, but can we try and limit the initial PR to the bare minimum to get the board and SoC going. So whatever basic init code and uart functioning. The other changes should than be split up into separate PRs.
> 
> For example, do we need the stimer driver for the board to boot and function?

service/ stuff is required for power management functionality which is added for EFR32BG22 SoC.

galak commented:
> > 
> 
> service/ stuff is required for power management functionality which is added for EFR32BG22 SoC.
> same with stimer, stimer is required for power manager.

Is that required to boot and have the SoC function?  If not, can we not add in a followup PR?

sateeshkotapati commented:
> > > 
> > 
> > 
> > service/ stuff is required for power management functionality which is added for EFR32BG22 SoC.
> > same with stimer, stimer is required for power manager.
> 
> Is that required to boot and have the SoC function? If not, can we not add in a followup PR?

board/configs is required to boot the SoC. We are trying to provide the functionality in Zephyr similar to Silabs sequence given in Simplicity Studio so it will be easier for users to migrate to latest SDKs, if required.

galak commented:
> > 
> 
> board/configs is required to boot the SoC. We are trying to provide the functionality in Zephyr similar to Silabs sequence given in Simplicity Studio so it will be easier for users to migrate to latest SDKs, if required.

What is required in board/configs to boot the SoC?  how is this different from the other silabs devices already supported in Zephyr?

sateeshkotapati commented:
> > > 
> > 
> > 
> > board/configs is required to boot the SoC. We are trying to provide the functionality in Zephyr similar to Silabs sequence given in Simplicity Studio so it will be easier for users to migrate to latest SDKs, if required.
> 
> What is required in board/configs to boot the SoC? how is this different from the other silabs devices already supported in Zephyr?

1. Initialization part is made similar to simplicity studio for easy migration to latest GSDKs.
2. There is no power manager initialization in already supported devices and some of initialization code required for EFR32BG22 SoC (like hfxo, dpll etc.,)  is missing in the existing code.

galak commented:
> > 
> 
>     1. Initialization part is made similar to simplicity studio for easy migration to latest GSDKs.
> 
>     2. There is no power manager initialization in already supported devices and some of initialization code required for EFR32BG22 SoC (like hfxo, dpll etc.,)  is missing in the existing code.

So I suggest seeing if you can support this SoC similar to how the other in tree SoCs are supported.  We can than look at the change to ALL of them to utilize the code from GSDK.

sateeshkotapati commented:
> > > 
> > 
> > 
> > ```
> > 1. Initialization part is made similar to simplicity studio for easy migration to latest GSDKs.
> > 
> > 2. There is no power manager initialization in already supported devices and some of initialization code required for EFR32BG22 SoC (like hfxo, dpll etc.,)  is missing in the existing code.
> > ```
> 
> So I suggest seeing if you can support this SoC similar to how the other in tree SoCs are supported. We can than look at the change to ALL of them to utilize the code from GSDK.

I am afraid we don't have any plan to change the already supported boards to utilize the code from GSDK at the moment and we don't want to break the functionality of already supported devices and impact its users.
We must need power manager functionality in EFR32BG22 to make it available as soon as possible for further contributions from Silabs.
 
The latest approach we have taken will not impact the already supported silabs devices and 
we will continue this new approach for silabs devices henceforth.

galak commented:
> I am afraid we don't have any plan to change the already supported boards to utilize the code from GSDK at the moment and we don't want to break the functionality of already supported devices and impact its users.
> We must need power manager functionality in EFR32BG22 to make it available as soon as possible for further contributions from Silabs.
> 
> The latest approach we have taken will not impact the already supported silabs devices and
> we will continue this new approach for silabs devices henceforth.

We need consistent support, so if you guys aren't interested in support / making the changes to the existing boards, than I ask you do things as the existing boards have done.

nashif commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,34 @@
+	/* Enable the board controller to be able to use the serial port */
+	vce_dev = device_get_binding(VCOM_ENABLE_GPIO_NAME);
+	if (!vce_dev) {
+		printk("Virtual COM Port Enable device was not found!\n");
```
use logging instead, i.e. LOG_ERR

chrta commented:
Removed my approval for now, because of the open discussion. I will re-review when the discussion is resolved.

WilliamGFish commented:
@sateeshkotapati
Has there been any progress with this? 
I am considering looking at the Thunderboard (sltb010a)

Billy..

sateeshkotapati commented:
> @sateeshkotapati
> Has there been any progress with this?
> I am considering looking at the Thunderboard (sltb010a)
> 
> Billy..

Hi Billy,

We are still having some internal discussions (at Silicon Labs) about the approach that we want to take on EFR32BG22.  We are planning to return to this pull request once those discussion are complete.  The timeline has not yet been finalized.

Regards,
Sateesh

WilliamGFish commented:
> > @sateeshkotapati
> > Has there been any progress with this?
> > I am considering looking at the Thunderboard (sltb010a)
> > Billy..
> 
> Hi Billy,
> 
> We are still having some internal discussions (at Silicon Labs) about the approach that we want to take on EFR32BG22. We are planning to return to this pull request once those discussion are complete. The timeline has not yet been finalized.
> 
> Regards,
> Sateesh

If you need assistance let me know.

Billy..

WilliamGFish commented:
> > > @sateeshkotapati
> > > Has there been any progress with this?
> > > I am considering looking at the Thunderboard (sltb010a)
> > > Billy..
> > 
> > 
> > Hi Billy,
> > We are still having some internal discussions (at Silicon Labs) about the approach that we want to take on EFR32BG22. We are planning to return to this pull request once those discussion are complete. The timeline has not yet been finalized.
> > Regards,
> > Sateesh
> 
> If you need assistance let me know.
> 
> Billy..

Any news ...?

Billy..

Delsian commented:
Do you have any updates? Is it possible to add module E104-BT53A3 (https://www.amazon.com/Bluetooth-Wireless-Interface-E104-BT53A3-Transceiver/dp/B09BCG9H3F) to the boards section?

sateeshkotapati commented:
Hello @galak, @chrta , and others,

After having multiple discussions internally about the approach to be taken going forward, came up with below conclusions:

1. We would like to go ahead with the latest approach that we have taken for EFR32BG22. This will ease the process of adding new silabs devices.
2. We will be adding support for the already available devices in Zephyr repo. However, it will be done incrementally, one device at a time and hoping we will be completing the process in next 6 months or so. Please note, the incremental addition will not impact the functionality of existing devices.
3. Modifying all the devices at once with latest approach would be too much effort and PR also will be difficult to review.
4. Also, we are volunteering to maintain the silabs devices officially.

Please let us know your opinion/comments ASAP so we can start the work immediately.  Thanks.

Regards,
Sateesh

chrta commented:
> 1. We would like to go ahead with the latest approach that we have taken for EFR32BG22. This will ease the process of adding new silabs devices.

For me that would be ok. But it would be important that anyone could add new silabs boards and socs.

So it is clearly documented which files are taken from gecko sdk, which are modified and/or generated. Also it must be clear how to generate these files and how to modify them. (this is probably important for the files in hal_silabs and not the files in this repo)

> 2. We will be adding support for the already available devices in Zephyr repo. However, it will be done incrementally, one device at a time and hoping we will be completing the process in next 6 months or so. Please note, the incremental addition will not impact the functionality of existing devices.
> 3. Modifying all the devices at once with latest approach would be too much effort and PR also will be difficult to review.

OK for me, but this should really be done within within some reasonable timeframe. There should only be one approach on how to add silabs devices in the future.

> 4. Also, we will be volunteering to maintain the silabs devices.

This would be great.

sateeshkotapati commented:
Please review this PR, Thanks.

gmarull commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,135 @@
+
+&usart0 {
+	current-speed = <115200>;
+	location-rx = <GECKO_LOCATION(0) GECKO_PORT_E GECKO_PIN(11)>;
+	location-tx = <GECKO_LOCATION(0) GECKO_PORT_E GECKO_PIN(10)>;
+	status = "okay";
+};
+
+&usart1 {
+	current-speed = <115200>;
+	location-rx = <GECKO_LOCATION(1) GECKO_PORT_A GECKO_PIN(6)>;
+	location-tx = <GECKO_LOCATION(1) GECKO_PORT_A GECKO_PIN(5)>;
```
a pinctrl driver needs to be implemented, please check https://github.com/zephyrproject-rtos/zephyr/issues/39740

chrta replied:
SiLabs MCUs (EFM32, EFR32) do not have a a dedicated pinctrl/pinmux peripheral. Each peripheral (e.g. this usart) has these location registers where the pin locations for this peripheral are configured.

As far as i understand there is no need for a pinctrl driver in that case.

gmarull replied:
The same applies for nRF chips, and pinctrl also applies. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pinctrl/pinctrl_nrf.c (drivers only supports UART for now)

sateeshkotapati replied:
> The same applies for nRF chips, and there's a pinctrl driver. https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pinctrl/pinctrl_nrf.c (drivers only supports UART for now)

We would like to take this up in a separate PR after this.

gmarull replied:
Custom pinctrl solutions are no longer allowed, new platforms need to make use of the API (unless you face severe limitations that require some API changes).

catch-twenty-two replied:
@sateeshkotapati @gmarull Are you referring to the fact that the board is using the "location-rx" prefix? Tmobile will also be submitting a dev board running the silabs chips set, that is also using the "location" labels. Please let me know how/if this will be changing.

gmarull replied:
Not really, what I mean is that pin control needs to go through the `pinctrl` API. For example, pin information is expected to be stored in `pinctrl-N` entries. There's a detailed guide here: https://docs.zephyrproject.org/latest/guides/pinctrl/index.html

sateeshkotapati committed:
pinctrl: Add Silabs Gecko implementation

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

gmarull commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -43,6 +43,31 @@
+			psels = <GECKO_PSEL(UART_TX, GECKO_PORT_E, GECKO_PIN(10))>,
+				<GECKO_PSEL(UART_RX, GECKO_PORT_E, GECKO_PIN(11))>,
+				<GECKO_LOC(UART_LOC, GECKO_LOCATION(0))>,
+				<GECKO_SPEED(UART_SPEED, 115200)>;
```
how is UART speed related to pinctrl? I think this should be a property of the `uart` device node itself.

sateeshkotapati replied:
Agreed, removed

gmarull commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -43,6 +43,31 @@
 
+&pinctrl {
+	compatible = "silabs,gecko-pinctrl";
+        /* configuration for uart0 device, default state */
+        usart0_default: usart0_default {
```
nit: please don't mix tabs/spaces

sateeshkotapati replied:
Done

gmarull commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -43,6 +43,31 @@
 };
 
+&pinctrl {
+	compatible = "silabs,gecko-pinctrl";
```
note: this is redundant, you're referencing an existing node (`&pinctrl`)

sateeshkotapati replied:
done

gmarull commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -43,6 +43,31 @@
 	};
 };
 
+&pinctrl {
```
by convention: define board pinctrl in `board-pinctrl.dtsi` and include it.

gmarull commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a_defconfig:
```diff
@@ -14,3 +14,4 @@ CONFIG_SOC_GECKO_EMU_DCDC=y
 CONFIG_SOC_GECKO_EMU_DCDC_MODE_ON=y
 CONFIG_HW_STACK_PROTECTION=y
 CONFIG_CMU_HFCLK_HFRCO=y
+CONFIG_PINCTRL=y
```
suggestion: default PINCTRL at SoC level if it's always required see e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/gigadevice/Kconfig.defconfig#L8

sateeshkotapati replied:
Currently BG22 has pinctrl support and it should be extended to other silabs devices in Zephyr codebase.

gmarull commented on drivers/pinctrl/Kconfig.gecko:
```diff
@@ -0,0 +1,10 @@
+	bool "Gecko pin controller driver"
+	depends on SOC_FAMILY_EXX32
+	select PINCTRL_STORE_REG
+	default y
```
please default based on `dt_compat_enabled`, see others

sateeshkotapati replied:
done

gmarull commented on dts/bindings/serial/silabs,gecko-usart.yaml:
```diff
@@ -38,3 +38,14 @@ properties:
       required: false
       description: CTS pin configuration defined as <location port pin>
+
+    pinctrl-0:
+      type: phandles
+      required: false
+      description: |
+        GPIO pin configuration for serial signals (RX, TX, RTS, CTS).  We expect
+        that the phandles will reference pinctrl nodes.
+
+        For example the USART1 would be
+           pinctrl-0 = <&usart1_tx_pb6 &usart1_rx_pb7>;
```
this is redundant property is already defined in `pinctrl-device`, if you want to document something about UART please just use the top-level description.

sateeshkotapati replied:
removed

gmarull commented on include/dt-bindings/pinctrl/gecko-pinctrl.h:
```diff
@@ -0,0 +1,104 @@
+ * @param port Port (0 or 1).
+ * @param pin Pin (0..31).
+ */
+#define GECKO_PSEL(fun, port, pin)						       \
+	(((port & GECKO_PORT_MSK) << GECKO_PORT_POS)		|			\
+	((pin & GECKO_PIN_MSK) << GECKO_PIN_POS) |	       \
+	((GECKO_FUN_ ## fun & GECKO_FUN_MSK) << GECKO_FUN_POS))
```
nit: please push all `\` to the same column

sateeshkotapati replied:
Done

gmarull reviewed (commented):
Thanks for the pinctrl bits, added some initial comments

sateeshkotapati commented:
I have incorporated the review comments.

gmarull commented:
> I have incorporated the review comments.

there are now 3 commits for pinctrl with the same title, please, squash them or split them in a meaningful way.

sateeshkotapati commented:
Hi Zephyr community, 
Please re-review this PR. I have incorporated all the previous review comments.
Our goal to add more Silabs devices is blocked on this PR. 
Thank you and appreciate your time. 

Regards,
Sateesh

henrikbrixandersen commented:
@sateeshkotapati Please rebase and fix the merge conflicts.

kmeinhar commented on drivers/pinctrl/pinctrl_gecko.c:
```diff
@@ -0,0 +1,98 @@
+{
+		struct soc_gpio_pin rxpin;
+		struct soc_gpio_pin txpin;
+		USART_TypeDef *base = (USART_TypeDef *)reg;
```
Pin control driver is USART specific, can this be made more generic? This way it could be used to assign other pin functions like the timer CC0 (PWM) output.

sateeshkotapati replied:
@kmeinhar, Gecko pin driver supports USART only for now. We would make it generic later once this PR is merged.

gmarull replied:
driver should be prepared to easily add more peripherals, see e.g. nRF one.

kmeinhar commented:
@sateeshkotapati Please consider adding Patch for PWM support. Verify that the register addresses and pins are correct, this patch was only verified on EFR32PG22-DK2503A.
[0001-drivers-Add-modify-PWM-support-for-EFR32BG22.patch.txt](https://github.com/zephyrproject-rtos/zephyr/files/8714882/0001-drivers-Add-modify-PWM-support-for-EFR32BG22.patch.txt)

sateeshkotapati commented:
> @sateeshkotapati Please rebase and fix the merge conflicts.

Done @henrikbrixandersen

carlescufi commented:
@gmarull please take another look

gmarull commented on drivers/counter/counter_gecko_stimer.c:
```diff
@@ -0,0 +1,296 @@
+static sl_sleeptimer_timer_handle_t alarm_timer[STIMER_ALARM_NUM];
+static sl_sleeptimer_timer_handle_t top_timer;
+
+#define DEV_NAME(dev) ((dev)->name)
+#define DEV_CFG(dev) \
+	((const struct counter_gecko_config *const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct counter_gecko_data *const)(dev)->data)
```
these _accessors_ were deleted from all drivers upstream, please delete them

sateeshkotapati replied:
Done

gmarull commented on drivers/serial/uart_gecko.c:
```diff
@@ -585,6 +612,29 @@ DT_INST_FOREACH_STATUS_OKAY(GECKO_UART_INIT)
 
+#ifdef CONFIG_PINCTRL
+#define GECKO_USART_INIT(idx)						       \
+	PINCTRL_DT_INST_DEFINE(idx);   \
```
nit: please align all `\`

gmarull commented on dts/arm/silabs/efr32bg22-pinctrl.dtsi:
```diff
@@ -0,0 +1,29 @@
+				<GECKO_LOC(UART_LOC, GECKO_LOCATION(1))>;
+		};
+	};
+};
```
missing EOF empty line

gmarull commented on dts/arm/silabs/efr32bg22-pinctrl.dtsi:
```diff
@@ -0,0 +1,29 @@
+	usart0_default: usart0_default {
+		group1 {
+			/* configure PE.11 as UART_RX and PE.10 as UART_TX */
+			psels = <GECKO_PSEL(UART_TX, GECKO_PORT_E, GECKO_PIN(10))>,
+				<GECKO_PSEL(UART_RX, GECKO_PORT_E, GECKO_PIN(11))>,
```
suggestion: for simplicity you could have `<GECKO_PSEL(UART_TX, E, 10)>`

gmarull commented on dts/bindings/pinctrl/silabs,gecko-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+          /* group 1 ('group1' name is arbitrary) */
+          group1 {
+            /* configure P0.1 as UART_TX and P0.2 as UART_RTS */
+            psels = <GECKO_PSEL(UART_TX, 0, 1)>, <GECKO_PSEL(UART_RTS, 0, 2)>;
```
this is not in line with the board above

gmarull commented on dts/bindings/pinctrl/silabs,gecko-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+    description: |
+      Silabs pin controller pin configuration group.
+    properties:
+      psels:
```
question: does silab use the 'psel' terminology, like nRF?

sateeshkotapati replied:
Yes

sateeshkotapati commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -10,6 +11,54 @@
+
+
+static const struct pm_state_info pm_min_residency[] =
+	PM_STATE_INFO_DT_ITEMS_LIST(DT_NODELABEL(cpu0));
```
@gmarull, It seems you are referring to earlier version of soc_power.c file
Latest file doesn't have these lines.
![image](https://user-images.githubusercontent.com/79091019/172316830-ad74738f-5376-4aa0-ba7e-0fc48f737a14.png)

sateeshkotapati commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,135 @@
+&usart1 {
+	current-speed = <115200>;
+	location-rx = <GECKO_LOCATION(1) GECKO_PORT_A GECKO_PIN(6)>;
+	location-tx = <GECKO_LOCATION(1) GECKO_PORT_A GECKO_PIN(5)>;
```
@gmarull, It seems you are referring to earlier version of this .dts file

sateeshkotapati commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -24,6 +73,54 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+
+	uint32_t BasePRI;
+
+	/* WorkAround: When this function is entered the Kernel has disabled
+	 * interrupts using BASEPRI register. This is incorrect as it prevents
+	 * waking up from any interrupt which priority is not 0. Work around the
+	 * issue using __set_BASEPRI(0) to remove the BASEPRI masking
+	 */
+
+	BasePRI = __get_BASEPRI();
+	__set_BASEPRI(0);
+
```
We require an event/interrupt should wakeup the device from sleep. So it is required to enable the interrupts before going to sleep.

sateeshkotapati committed:
boards: Add support for Silicon Labs' EFR32BG22-SLTB010A board

Incorporated review comments.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
drivers: gecko: Add/modify support for EFR32BG22 drivers

Incorporated review comments.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
drivers: gecko: Add/modify support for EFR32BG22 drivers

Incorporated review comments.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
soc: silabs_exx32: Add support for SiLabs EFR32BG22 SoC

Incorporated review comments.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati committed:
soc: silabs_exx32: Add support for SiLabs EFR32BG22 SoC

Incorporated review comments.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>

sateeshkotapati commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,37 @@
+#include <sys/printk.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(efr32bg_sltb010a, LOG_LEVEL_DBG);
```
Done

sateeshkotapati commented on boards/arm/efr32bg_sltb010a/board.c:
```diff
@@ -0,0 +1,37 @@
+	gpio_pin_configure(vce_dev, VCOM_ENABLE_GPIO_PIN, GPIO_OUTPUT_HIGH);
+
+	return 0;
+}
```
Done

sateeshkotapati commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,135 @@
+	leds {
+		compatible = "gpio-leds";
+		led0: led_0 {
+			gpios = <&gpiob 0 0>;
```
Done

sateeshkotapati commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,135 @@
+	buttons {
+		compatible = "gpio-keys";
+		button0: button_0 {
+			/* gpio flags need validation */
```
Removed

sateeshkotapati commented on boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts:
```diff
@@ -0,0 +1,135 @@
+
+&gpiof {
+	status = "okay";
+};
```
Removed unused gpio ports

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/soc.h:
```diff
@@ -0,0 +1,29 @@
+#include "../common/soc_gpio.h"
+
+/* Add include for DTS generated information */
+#include <devicetree.h>
```
removed

sateeshkotapati commented on soc/arm/silabs_exx32/efr32bg22/soc_pinmap.h:
```diff
@@ -0,0 +1,34 @@
+#define PIN_SWO { gpioPortC, 11, gpioModePushPull, 1 }
+#elif (SWO_LOCATION >= 4)
+#error ("Invalid SWO pin location")
+#endif
```
This file is not required, removed.

sateeshkotapati commented on drivers/pinctrl/pinctrl_gecko.c:
```diff
@@ -0,0 +1,98 @@
+				case GECKO_FUN_UART_TX:
+					txpin.port = GECKO_GET_PORT(pins[i]);
+					txpin.pin = GECKO_GET_PIN(pins[i]);
+					txpin.mode = gpioModePushPull;
```
pin mode is not a Boolean unlike drive-push-pull and it is difficult to cover from standard property.

sateeshkotapati commented on soc/arm/silabs_exx32/common/soc_power.c:
```diff
@@ -50,11 +50,12 @@ static void sli_os_schedule_wakeup(uint32_t expected_idletime_in_os_ticks);
 static const struct pm_state_info pm_min_residency[] =
-	PM_STATE_INFO_DT_ITEMS_LIST(DT_NODELABEL(cpu0));
+	PM_STATE_INFO_LIST_FROM_DT_CPU(DT_NODELABEL(cpu0));
+struct pm_state_info pm_state_active = {PM_STATE_ACTIVE, 0, 0, 0};
```
These changes are not part of pinctrl, these changes were pulled from master while pushing pinctrl changes.

smalae committed:
Updated formatting for checkpatch

smalae committed:
Updated formatting for checkpatch

sateeshkotapati committed:
tests: kernel: profiling: Exclude EFR32BG22-SLTB010A board

This commit excludes EFR32BG22-SLTB010A (a.k.a Thunderboard EFR32BG22)
from running this testcase.

Signed-off-by: Sateesh Kotapati <Sateesh.Kotapati@silabs.com>
